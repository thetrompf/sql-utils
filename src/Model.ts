import { ColumnChanged } from './ColumnChanged';
import { ColumnUpdate } from './ColumnUpdate';
import { ForeignColumnChanged } from './ForeignColumnChanged';
import { ForeignColumnUpdate } from './ForeignColumnUpdate';
import { ColumnSelect } from './ColumnSelect';
import { ForeignColumnSelect } from './ForeignColumnSelect';
import { ColumnMap } from './ColumnMap';
import { changed } from './Changed';
import { update } from './Update';
import { foreignChanged } from './ForeignChanged';
import { foreignUpdate } from './ForeignUpdate';
import { select } from './Select';
import { foreignSelect } from './ForeignSelect';
import { foreignSelectNoPrefix } from './ForeignSelectNoPrefix';
import { selectNoPrefix } from './SelectNoPrefix';

export type Model<TData> = ModelInterface<TData> & TData;

interface ModelInterface<TData> {
    columnChanged(paramName: string, tableAlias: string): ColumnChanged;
    columnUpdate(paramName: string): ColumnUpdate;
    foreignColumnChanged(foreignTableName: string, tableAlias: string, paramName: string): ForeignColumnChanged;
    foreignColumnUpdate(foreignTableName: string, paramName: string): ForeignColumnUpdate;
    updateWithData(data: TData): void;
}

export interface ModelConstructor<TData> {
    readonly TypeName: string;
    new (data: TData): Model<TData>;

    columnSelect(tableAlias: string, prefix?: string): ColumnSelect<TData>;
    foreignColumnSelect(foreignTableName: string, tableAlias: string, prefix?: string): ForeignColumnSelect<TData>;
    foreignInsertReturning(foreignTableName: string): ForeignColumnSelect<TData>;
    foreignOutputReselect(foreignTableName: string, tableAlias: string, prefix?: string): ForeignColumnSelect<TData>;
    insertReturning(): ColumnSelect<TData>;
    outputReselect(tableAlias: string, prefix?: string): ColumnSelect<TData>;
}

export function baseModelGenerator<TData>(name: string, columnMap: ColumnMap) {
    const reselectColumnMap = Object.keys(columnMap).reduce(
        (carry, columnName) => {
            const mapped = columnMap[columnName];
            carry[columnName] = {
                columnName: columnName,
                foreignTableName: typeof mapped === 'string' ? undefined : mapped.foreignTableName,
            };
            return carry;
        },
        {} as ColumnMap,
    );
    Object.freeze(reselectColumnMap);

    abstract class AbstractBaseModel implements ModelInterface<TData> {
        public static columnSelect(tableAlias: string, prefix?: string | undefined): ColumnSelect<TData> {
            return select<TData>(tableAlias, columnMap, prefix);
        }

        public static foreignColumnSelect(
            foreignTableName: string,
            tableAlias: string,
            prefix?: string | undefined,
        ): ForeignColumnSelect<TData> {
            return foreignSelect<TData>(foreignTableName, tableAlias, columnMap, prefix);
        }

        public static foreignInsertReturning(foreignTableName: string): ForeignColumnSelect<TData> {
            return foreignSelectNoPrefix<TData>(foreignTableName, columnMap);
        }

        public static foreignOutputReselect(
            foreignTableName: string,
            tableAlias: string,
            prefix?: string | undefined,
        ): ForeignColumnSelect<TData> {
            return foreignSelect<TData>(foreignTableName, tableAlias, columnMap, prefix);
        }

        public static insertReturning(): ColumnSelect<TData> {
            return selectNoPrefix<TData>(columnMap);
        }

        public static outputReselect(tableAlias: string, prefix?: string | undefined): ColumnSelect<TData> {
            return select<TData>(tableAlias, columnMap, prefix);
        }

        public columnChanged(paramName: string, tableAlias: string): ColumnChanged {
            return changed(tableAlias, columnMap, paramName, this);
        }

        public columnUpdate(paramName: string): ColumnUpdate {
            return update(columnMap, paramName, this);
        }

        public foreignColumnChanged(
            foreignTableName: string,
            tableAlias: string,
            paramName: string,
        ): ForeignColumnChanged {
            return foreignChanged(foreignTableName, tableAlias, columnMap, paramName, this);
        }

        public foreignColumnUpdate(foreignTableName: string, paramName: string): ForeignColumnUpdate {
            return foreignUpdate(foreignTableName, columnMap, paramName, this);
        }

        public updateWithData(data: TData): void {
            Object.assign(this, data);
        }

        public static get TypeName() {
            return name;
        }
    }

    return AbstractBaseModel;
}
