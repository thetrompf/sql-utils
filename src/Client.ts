interface QueryResultBase {
    command: string;
    rowCount: number;
}

export interface QueryResult extends QueryResultBase {
    rows: any[];
}

export interface QueryArrayResult extends QueryResultBase {
    rows: any[][];
}

export interface Client<T> {
    readonly transactionIsolationLevel: number;
    beginTransaction(): Promise<void>;
    close(): void;
    closeAsync(): Promise<void>;
    commit(): Promise<void>;
    getTransactionUniqueNumber(): number;
    getUnderlyingClient(): T;
    query(sql: string, parameters?: any[], name?: string): Promise<QueryResult>;
    rollback(): Promise<void>;
}
