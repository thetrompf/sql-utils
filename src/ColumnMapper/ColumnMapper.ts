import { InputColumnMapper } from './InputColumnMapper';
import { OutputColumnMapper } from './OutputColumnMapper';

export interface ColumnMapper {
    columnName: string;
    foreignTableName?: string;
    input?: InputColumnMapper;
    output?: OutputColumnMapper;
    readOnly?: boolean;
}
