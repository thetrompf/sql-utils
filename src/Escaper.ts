export interface Escaper {
    identifier(id: string): string;
}
