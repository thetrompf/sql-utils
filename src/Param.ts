import { ParamInput } from './ParamInput';

export function param(name: string, value: any): ParamInput {
    return new ParamInput(name, value);
}
