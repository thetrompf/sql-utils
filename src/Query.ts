import { TemplateInput } from './TemplateInput';
import { Template } from './Template';

export function Query<T = any>(parts: TemplateStringsArray, ...input: TemplateInput<T>[]): Template<T> {
    return new Template<T>(parts, input);
}
