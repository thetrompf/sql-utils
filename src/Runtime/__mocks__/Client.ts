import { Client, QueryResult } from '../Client';

export class MockClient implements Client<any> {
    public readonly transactionNestingLevel: number = 0;
    public beginTransaction(): Promise<void> {
        return Promise.resolve();
    }
    public close(): void {
        return;
    }
    public closeAsync(): Promise<void> {
        return Promise.resolve();
    }
    public commit(): Promise<void> {
        return Promise.resolve();
    }
    public getTransactionUniqueNumber(): number {
        return 0;
    }
    public getUnderlyingClient() {
        return null;
    }
    private result: QueryResult | null = null;
    public setResult(result: QueryResult | null) {
        this.result = result;
    }
    public query(_sql: string, _parameters?: any[] | undefined, _name?: string | undefined): Promise<QueryResult> {
        if (this.result == null) {
            throw new Error('QueryResult is not set');
        }
        return Promise.resolve(this.result);
    }
    public rollback(): Promise<void> {
        return Promise.resolve();
    }
}
