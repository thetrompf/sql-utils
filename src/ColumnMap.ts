import { ColumnMapper } from './ColumnMapper';

export interface ColumnMap {
    [column: string]: string | ColumnMapper;
}
