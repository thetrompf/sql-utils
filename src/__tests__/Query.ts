import { ColumnMap } from '../ColumnMapper/ColumnMap';
import { baseModelGenerator } from '../Model';
import { Query as sql } from '../Query';
import { param } from '../Select/Param';
import { MockClient } from '../Runtime/__mocks__/Client';
import { MockPlatform } from '../Runtime/__mocks__/Platform';
import { mockEscaper } from '../Runtime/__mocks__/Escaper';

const client = new MockClient();
const platform = new MockPlatform(client, mockEscaper);

describe('Query', () => {
    beforeEach(() => {
        platform.client.setResult(null);
    });

    test('Simple model select', async () => {
        const userColumnMap: ColumnMap = {
            id: 'id',
            name: 'name',
            email: 'email',
        };

        interface UserData {
            id: number;
            name: string;
            email: string;
        }

        const UserModelBase = baseModelGenerator<UserData>('User', userColumnMap);
        class UserModel extends UserModelBase {
            public id: number;
            public name: string;
            public email: string;

            public constructor(data: UserData) {
                super();
                this.id = data.id;
                this.name = data.name;
                this.email = data.email;
            }
        }

        const idParam = param('id', 1);
        const template = sql`
            SELECT
                ${UserModel.columnSelect('u')}
            FROM
                public.users u
            WHERE
                u.id = ${idParam}
        `;

        expect(template.getSql(platform)).toMatchSnapshot();

        client.setResult({
            command: 'SELECT',
            rowCount: 1,
            rows: [
                {
                    id: 1,
                    name: 'Test person',
                    email: 'test@example.com',
                },
            ],
        });

        const userResult = await template.singleResult(platform);
        const user = new UserModel(userResult);

        expect(user).toBeInstanceOf(UserModel);
        expect(user.id).toBe(1);
        expect(user.name).toBe('Test person');
        expect(user.email).toBe('test@example.com');
    });
});
