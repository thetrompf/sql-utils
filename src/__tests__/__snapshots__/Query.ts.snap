// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query Simple model select 1`] = `
"
            SELECT
                u.id as id, u.name as name, u.email as email
            FROM
                public.users u
            WHERE
                u.id = $1
        "
`;
